### PR TITLE
SDR concat fix const signature

### LINKED ISSUE #579: SDR concatenate problem?
I have compiling error in Ubuntu by using a vector for concatenating 2 SDRs together.
Can someone help me?
here is my code:
```
    SDR A({10});
    SDR B({10});
    SDR C({20});        
    SDR D({20});
    A.randomize( 0.30f );
    B.randomize( 0.70f );
    
    D.concatenate({&A, &B});	// OK

    std::vector<SDR*> AB;
    AB.push_back(&A);
    AB.push_back(&B);
    C.concatenate(AB);	// why failed???
```

### PR DESCRIPTION
**Pull request by breznak**
Fixes #579

### CODE REVIEW

File: src/test/unit/types/SdrTest.cpp [hunk 1]
```diff
@@ -673,12 +673,25 @@ TEST(SdrTest, TestConcatenation) {
...
     ASSERT_EQ( C.getSum(), 10u );
-    D.concatenate({&A, &B});
+
+    //concat a vector
+    EXPECT_NO_THROW(D.concatenate({&A, &B}));
     ASSERT_EQ( D.getSum(), 10u );
+
+    //concat explicit vector
+    std::vector<const SDR*> AB;
+    AB.push_back(&A);
```
Comment (breznak) [inline src/test/unit/types/SdrTest.cpp]: adds test case for explicit vector for concat

**Comment by breznak** [comment]
OT: is there a specific reason why we're using a vector of pointers, rather than &, or even `vector<SDR>&` ?

**Comment by ctrl-z-9000-times** [comment]
> OT: is there a specific reason why we're using a vector of pointers, rather than &, or even vector<SDR>& ?

C++ does not like `vector<&>`, the compile fails

I think `vector<SDR> &` would incur an extra data copy, when the user moves their data from their local variable into the input list:
```C++
SDR A = SDR(123);
SDR B = SDR(123);
vector<SDR> concat_inputs({A, B}); // I think this line copies all of the data?
SDR C = SDR(246).concatenate( (vector<SDR>&) concat_inputs );
```

**Comment by breznak** [comment]
> C++ does not like vector<&>, the compile fails

yea, I've tried. Not sure why though. 

> SDR B = SDR(123);
vector<SDR> concat_inputs({A, B}); // I think this line copies all of the data?

probably. There're some mitigations 
https://stackoverflow.com/questions/45143043/why-does-stdvector-enforce-copy-on-initialization

not sure if worth it. The idea was to avoid pointers, and just require vector of SDRs (if SDR& do not work as elements of a vector)

**Comment by ctrl-z-9000-times** [comment]
> The idea was to avoid pointers

My solution to that would be to provide a few different overloads for concatenating different numbers of inputs.

**Comment by Thanh-Binh** [comment]
Why did you not implement concatenate() like push_back() for vector? I mean:
SDR C;
C.concatenate(A);
C.concatenate(B);

**Comment by breznak** [comment]
> implement concatenate() like push_back() 

I would like that syntax. I think one of the reasons is SDR cannot change its size, so we'd return a new SDR on each concat.

**Pull request merged**